programs/ziti-edge-tunnel/netif_driver/windows/tun.c

```diff
@@ -0,0 +1,262 @@
+        X(WintunDeletePoolDriver, WINTUN_DELETE_POOL_DRIVER_FUNC) ||
+        X(WintunEnumAdapters, WINTUN_ENUM_ADAPTERS_FUNC) ||
+        X(WintunFreeAdapter, WINTUN_FREE_ADAPTER_FUNC) ||
+        X(WintunOpenAdapter, WINTUN_OPEN_ADAPTER_FUNC) ||
+        X(WintunGetAdapterLUID, WINTUN_GET_ADAPTER_LUID_FUNC) ||
+        X(WintunGetAdapterName, WINTUN_GET_ADAPTER_NAME_FUNC) ||
+        X(WintunSetAdapterName, WINTUN_SET_ADAPTER_NAME_FUNC) ||
+        X(WintunGetRunningDriverVersion, WINTUN_GET_RUNNING_DRIVER_VERSION_FUNC) ||
+        X(WintunSetLogger, WINTUN_SET_LOGGER_FUNC) ||
+        X(WintunStartSession, WINTUN_START_SESSION_FUNC) ||
+        X(WintunEndSession, WINTUN_END_SESSION_FUNC) ||
+        X(WintunGetReadWaitEvent, WINTUN_GET_READ_WAIT_EVENT_FUNC) ||
+    X(WintunReceivePacket, WINTUN_RECEIVE_PACKET_FUNC) ||
+    X(WintunReleaseReceivePacket, WINTUN_RELEASE_RECEIVE_PACKET_FUNC) ||
+    X(WintunAllocateSendPacket, WINTUN_ALLOCATE_SEND_PACKET_FUNC) ||
+    X(WintunSendPacket, WINTUN_SEND_PACKET_FUNC))
+#undef X
+    {
+        DWORD LastError = GetLastError();
+        FreeLibrary(Wintun);
+        SetLastError(LastError);
+        Wintun = NULL;
+    }
+
+    WINTUN = Wintun;
+}
+
+netif_driver tun_open(struct uv_loop_s *loop, uint32_t tun_ip, uint32_t dns_ip, const char *cidr, char *error, size_t error_len) {
+    if (error != NULL) {
+        memset(error, 0, error_len * sizeof(char));
+    }
+
+    uv_once(&wintunInit, InitializeWintun);
+    if (WINTUN == NULL) {
+        strcpy_s(error, error_len, "Failed to load wintun.dll");
+        return NULL;
+    }
+
+    struct netif_handle_s *tun = calloc(1, sizeof(struct netif_handle_s));
+    if (tun == NULL) {
+        if (error != NULL) {
+            snprintf(error, error_len, "failed to allocate tun");
+        }
+        return NULL;
+    }
+
+    GUID adapterGuid;
+    IIDFromString(L"2cbfd72d-370c-43b0-b0cd-c8f092a7e134", &adapterGuid);
+    tun->adapter = WintunOpenAdapter(L"Ziti", L"tun0");
+    if (!tun->adapter) {
+        tun->adapter = WintunCreateAdapter(L"Ziti", L"tun0", &adapterGuid, NULL);
+        if (!tun->adapter) {
+            DWORD err = GetLastError();
+            snprintf(error, error_len, "Failed to create adapter: %d", err);
+            return NULL;
+        }
+    }
+
+    DWORD Version = WintunGetRunningDriverVersion();
+    ZITI_LOG(INFO, "Wintun v%u.%u loaded", (Version >> 16) & 0xff, (Version >> 0) & 0xff);
+
+    MIB_UNICASTIPADDRESS_ROW AddressRow;
+    InitializeUnicastIpAddressEntry(&AddressRow);
+    WintunGetAdapterLUID(tun->adapter, &AddressRow.InterfaceLuid);
+    AddressRow.Address.Ipv4.sin_family = AF_INET;
+    AddressRow.Address.Ipv4.sin_addr.S_un.S_addr = tun_ip;
+    AddressRow.OnLinkPrefixLength = 16;
+    DWORD err = CreateUnicastIpAddressEntry(&AddressRow);
+    if (err != ERROR_SUCCESS && err != ERROR_OBJECT_ALREADY_EXISTS)
+    {
+        ZITI_LOG(ERROR, "Failed to set IP address: %d", err);
```

Comment:
why the ZITI_LOG here but above there was no ziti_log? seems like this  would set the error/error_len out params instead of logging? or at least all the usage would be consistent, one way or the other?